File: csharp/927.three-equal-parts.cs
/*
 * @lc app=leetcode id=927 lang=csharp
 *
 * [927] Three Equal Parts
 */

using System.Linq;

// @lc code=start
public class Solution {
    public int[] ThreeEqualParts(int[] A) {
        int n = A.Length, count = A.Count(x => x == 1);
        if (count % 3 != 0) {
            return new int[] { -1, -1 };
        }
        if (count == 0) {
            return new int[] { 0, n - 1 };
        }
        count /= 3;
        int start = -1, mid, end, curr = 0;
        while (curr == 0) {
            ++start;
            if (A[start] == 1) {
                ++curr;
            }
        }
        mid = start;
        while (curr <= count) {
            ++mid;
            if (A[mid] == 1) {
                ++curr;
            }
        }
        end = mid;
        while (curr <= 2 * count) {
            ++end;
            if (A[end] == 1) {
                ++curr;
            }
        }
        while (end < n) {
            int x = A[start++];
            if (x != A[mid++] || x != A[end++]) {
                return new int[] { -1, -1 };
            }
        }
        if (end == n) {
            return new int[] { start - 1, mid };
        }
        return new int[] { -1, -1 };
    }
}
// @lc code=end

File: java/927.three-equal-parts.java
/*
 * @lc app=leetcode id=927 lang=java
 *
 * [927] Three Equal Parts
 */

// @lc code=start
class Solution {
    public int[] threeEqualParts(int[] arr) {
        int n = arr.length, p1 = -1, p2, p3, acc = 0, count = 0;
        for (int i = 0; i < n; ++i) {
            if (arr[i] == 1) {
                ++count;
            }
        }
        if (count == 0) {
            return new int[] { 0, n - 1 };
        }
        if (count % 3 != 0) {
            return new int[] { -1, -1 };
        }
        count /= 3;
        while (acc == 0) {
            ++p1;
            if (arr[p1] == 1) {
                ++acc;
            }
        }
        p2 = p1;
        while (acc <= count) {
            ++p2;
            if (arr[p2] == 1) {
                ++acc;
            }
        }
        p3 = p2;
        while (acc <= 2 * count) {
            ++p3;
            if (arr[p3] == 1) {
                ++acc;
            }
        }
        while (p3 < n) {
            int v = arr[p3++];
            if (v != arr[p1++] || v != arr[p2++]) {
                return new int[] { -1, -1 };
            }
        }
        return new int[] { p1 - 1, p2 };
    }
}
// @lc code=end

